### PR TITLE
feat(ws): add multi-client awareness protocol (#10)

### LINKED ISSUE #10: Desktop/mobile session handoff
## Summary
Allow users to seamlessly swap between desktop and mobile for the same Claude Code session. Since the session lives in tmux, the underlying terminal is persistent — we just need to handle multiple clients viewing/controlling the same session.

## Current Behavior
- Only one client connects at a time via WebSocket
- Reconnecting from a different device works but there's no handoff UX

## Desired Behavior
- User can open Chroxy on desktop browser OR mobile app
- Both can view the session (read-only for one, or full duplex)
- Clean handoff: "Another device connected" notification
- Session state (chat history, terminal buffer) syncs on connect

## Technical Considerations
- tmux already supports multiple clients (`tmux attach`)
- WsServer currently supports multiple connections — need to decide on control semantics
- May need a lightweight web client for desktop access
- Related to #3 (multi-tab support)

## Related
- Depends on better output parsing or a structured API from Claude Code (see investigation notes)

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

- Add client tracking so connected devices know about each other
- Enhanced auth flow: clients send `deviceInfo`, receive `connectedClients` list in `auth_ok`
- New WS messages: `client_joined`, `client_left`, `primary_changed` (last-writer-wins per session)
- SettingsBar shows device count badge and expanded device list with "(this device)" indicator

Closes #10

## Changes

| File | What changed |
|------|-------------|
| `packages/server/src/ws-server.js` | Client tracking, deviceInfo extraction, join/leave broadcasts, primary tracking |
| `packages/app/src/store/connection.ts` | ConnectedClient type, device ID persistence, join/leave/primary handlers |
| `packages/app/src/components/SettingsBar.tsx` | Device count badge + expanded "Connected Devices" section |
| `packages/app/src/screens/SessionScreen.tsx` | Wire connectedClients prop to SettingsBar |
| `packages/server/tests/ws-server.test.js` | 9 new tests (multi-client awareness + primary tracking) |
| `packages/app/src/__tests__/store/connection.test.ts` | 6 new tests (connectedClients state management) |
| `CLAUDE.md` | Updated WS protocol docs with new message types |

## Protocol Details

**Enhanced messages:**
- `auth` → optional `deviceInfo: { deviceId, deviceName, deviceType, platform }`
- `auth_ok` → `clientId` + `connectedClients` array

**New messages (Server → Client):**
- `client_joined` — broadcast to other clients when a new client authenticates
- `client_left` — broadcast when a client disconnects
- `primary_changed` — broadcast when last-writer changes for a session

**Edge cases handled:**
- Old clients without `deviceInfo` get safe defaults
- Primary cleared on disconnect, re-established on next input
- Backward compatible — no breaking changes to existing protocol

## Test Plan

- [x] 457 server tests pass (9 new)
- [x] 43 app tests pass (6 new)
- [x] TypeScript compilation clean
- [ ] Manual: connect two devices, verify badges + system messages

### CODE REVIEW

**Comment by blamechris** [comment]
## Agent Review — PR #411

**Verdict: APPROVE with suggestions**

---

### Summary

Solid implementation of multi-client awareness for the WebSocket protocol. The server tracks connected clients with device metadata, broadcasts join/leave events, and implements last-writer-wins primary tracking per session. The app side handles the new messages with proper type safety, system chat messages, and a clean SettingsBar badge/detail UI. Backward compatible — old clients without `deviceInfo` get safe defaults. Good test coverage (9 server + 6 app tests).

### What Looks Good

- **Backward compatibility**: Old clients without `deviceInfo` gracefully default to `null`/`'unknown'` — no breaking changes
- **Defensive parsing**: Both server and client validate all incoming fields with type guards and fallback defaults
- **Proper cleanup on departure**: `_handleClientDeparture` clears primary tracking AND broadcasts `client_left` — covers both normal close and ping-timeout termination
- **Idempotent primary tracking**: `_updatePrimary` short-circuits when client is already primary — avoids redundant broadcasts
- **Auth guard on departure**: Unauthenticated clients do not trigger join/leave broadcasts (tested)
- **Type safety**: `ConnectedClient` interface is well-defined, `isSelf` detection works via `clientId` match, union types for `deviceType`
- **Test coverage**: Good edge cases — old clients, duplicate primary, disconnect cleanup, unauthenticated clients

### Suggestions

#### S1: `_primaryClients` not cleared in `close()`

In `ws-server.js` `close()` (line ~1466), `_pendingPermissions` is properly cleared but `_primaryClients` is not. While not a live bug (the server is shutting down), it breaks the cleanup symmetry pattern established by the rest of the method.

```js
// After this._pendingPermissions.clear():
this._primaryClients.clear()
```

#### S2: Primary tracking missing in legacy single-session modes

`_updatePrimary()` is only called from `_handleSessionMessage` (multi-session mode, line 497). The legacy `_handleCliMessage` (line 756) and `_handlePtyMessage` (line 827) do not call it. If a user runs the server in single-CLI or PTY mode with multiple connected devices, `primary_changed` will never fire. This is fine if those modes are deprecated, but worth documenting or adding the call for consistency.

#### S3: Auth message delayed by async `getDeviceId()` — potential race

In `connection.ts` (line 738), `socket.onopen` fires and then awaits `getDeviceId()` before sending the `auth` message. If the SecureStore read takes time, there's a window where the server's 10-second auth timeout is counting down but no auth message has been sent yet. In practice, SecureStore is fast and the cached path skips the await entirely on subsequent connects. The `readyState` check is a good guard. Low risk but worth noting.

#### S4: `myClientId` not persisted in store for future use

The `clientId` from `auth_ok` is parsed (line 772) but only used locally for `isSelf` detection. It is not stored in Zustand state. If future features need to know "which client am I?" (e.g., showing primary badge on self), you will need to add a `myClientId` field to the store. Consider adding it now to avoid re-plumbing later.

#### S5: `client_joined` can add duplicates on rapid reconnect

If a client disconnects and reconnects fast enough that the `client_left` for the old connection hasn't been processed yet by another client, the `client_joined` handler (line 1384) appends without checking for duplicate `clientId`. Since the server assigns new IDs per connection this should not happen in practice, but a dedup filter would be defensive:

```ts
connectedClients: [...state.connectedClients.filter(c => c.clientId !== newClient.clientId), newClient],
```

#### S6: Minor — "cannot log after tests are done" warning

The test cleanup for the primary client tracking tests produces a harmless Jest warning about `console.log` after test completion. This is because `ws.close()` is not fully awaited before the test exits. Not a correctness issue but shows up in CI output.

### Issues

No critical or blocking issues found.

### Follow-up Issues

See linked issues below for deferred items.

---

*Reviewed by agent-review at commit a5e1a42*

File: packages/app/src/store/connection.ts [hunk 1]
```diff
@@ -1292,6 +1370,75 @@ export const useConnectionStore = create<ConnectionState>((set, get) => ({
...
+            timestamp: Date.now(),
+          };
+          const joinActiveId = get().activeSessionId;
+          if (joinActiveId && get().sessionStates[joinActiveId]) {
+            updateActiveSession((ss) => ({
+              messages: [...ss.messages, joinMsg],
+            }));
+          } else {
+            get().addMessage(joinMsg);
+          }
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/app/src/store/connection.ts]: Valid point — the app-side tests verify store shape via `setState` but don't exercise the WS message handler path. Created follow-up issue: https://github.com/blamechris/chroxy/issues/416

Note that the server-side tests in `ws-server.test.js` do have full integration coverage for these events (broadcasting `client_joined`/`client_left`/`primary_changed` over real WebSocket connections). The gap is specifically on the app-side handler that parses these messages and updates Zustand state.
